retrievalmarket/server/gsunpaidretrieval.go

```diff
@@ -0,0 +1,475 @@
+		g.publishMktsEvent(retrievalmarket.ProviderEventCancelComplete, *state.mkts)
+
+		g.untrackTransfer(p, state.cs.transferID)
+	})
+}
+
+func (g *GraphsyncUnpaidRetrieval) RegisterBlockSentListener(listener graphsync.OnBlockSentListener) graphsync.UnregisterHookFunc {
+	return g.GraphExchange.RegisterBlockSentListener(func(p peer.ID, request graphsync.RequestData, block graphsync.BlockData) {
+		_, state, intercept := g.isRequestForActiveUnpaidRetrieval(p, request)
+		if !intercept {
+			listener(p, request, block)
+			return
+		}
+
+		// When a data transfer is restarted, the requester sends a list of CIDs
+		// that it already has. Graphsync calls the sent hook for all blocks even
+		// if they are in the list (meaning, they aren't actually sent over the
+		// wire). So here we check if the block was actually sent
+		// over the wire before firing the data sent event.
+		if block.BlockSizeOnWire() == 0 {
+			return
+		}
+
+		// Fire block sent event
+		state.cs.sent += block.BlockSizeOnWire()
+		g.publishDTEvent(datatransfer.DataSent, "", state.cs)
+		state.mkts.TotalSent += block.BlockSizeOnWire()
+	})
+}
+
+func (g *GraphsyncUnpaidRetrieval) RegisterNetworkErrorListener(listener graphsync.OnNetworkErrorListener) graphsync.UnregisterHookFunc {
+	return g.GraphExchange.RegisterNetworkErrorListener(func(p peer.ID, request graphsync.RequestData, err error) {
+		_, state, intercept := g.isRequestForActiveUnpaidRetrieval(p, request)
+		if !intercept {
+			listener(p, request, err)
+			return
+		}
+
+		// Don't set the channel state to errored, as a network error is not
+		// necessarily fatal
+		// TODO: If the client always sends a restart message after a network
+		// disconnect, then maybe a network error *should* be fatal
+		g.publishDTEvent(datatransfer.Error, fmt.Sprintf("network error: %s", err), state.cs)
+	})
+}
+
+func (g *GraphsyncUnpaidRetrieval) failTransfer(state *retrievalState, err error) {
+	state.cs.status = datatransfer.Failed
+	state.cs.message = err.Error()
+	g.publishDTEvent(datatransfer.Error, err.Error(), state.cs)
+	state.mkts.Status = retrievalmarket.DealStatusErrored
+	g.publishMktsEvent(retrievalmarket.ProviderEventDataTransferError, *state.mkts)
+
+	g.untrackTransfer(state.cs.recipient, state.cs.transferID)
+	log.Infow("transfer failed", "transfer id", state.cs.transferID, "peer", state.cs.recipient, "err", err)
+}
+
+func (g *GraphsyncUnpaidRetrieval) decodeVoucher(request datatransfer.Request, registry *registry.Registry) (datatransfer.Voucher, error) {
+	vtypStr := request.VoucherType()
+	decoder, has := registry.Decoder(vtypStr)
+	if !has {
+		return nil, xerrors.Errorf("unknown voucher type: %s", vtypStr)
+	}
+	encodable, err := request.Voucher(decoder)
+	if err != nil {
+		return nil, err
+	}
+	return encodable.(datatransfer.Registerable), nil
+}
+
+func (g *GraphsyncUnpaidRetrieval) isRequestForActiveUnpaidRetrieval(p peer.ID, request graphsync.RequestData) (datatransfer.Request, *retrievalState, bool) {
```

Comment:
as I read this, this only returns true if a retrieval was previously tracked with trackTransfer? so this is called for all the hooks after IncomingRequestQueued?